File: src/main/java/com/example/TelegramBot/MyBot.java
package com.example.TelegramBot;

import org.telegram.telegrambots.bots.TelegramLongPollingBot;
import org.telegram.telegrambots.meta.api.methods.send.SendDocument;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.methods.send.SendPhoto;
import org.telegram.telegrambots.meta.api.objects.InputFile;
import org.telegram.telegrambots.meta.api.objects.Update;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.ReplyKeyboardMarkup;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.KeyboardRow;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.InlineKeyboardButton;

import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public class MyBot extends TelegramLongPollingBot {
    private final MyMemoryTranslator translator = new MyMemoryTranslator();
    private String currentFromLang = "en";
    private String currentToLang = "es";

    @Override
    public void onUpdateReceived(Update update) {
        if (update.hasMessage() && update.getMessage().hasText()) {
            String text = update.getMessage().getText();
            long chatId = update.getMessage().getChatId();
            String translated;
            try {
                translated = translator.translate(text, currentFromLang, currentToLang);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }

            if (text.startsWith("/start")) {
                sendMainMenu(chatId);
            }
            else if (text.startsWith("/from ")) {
                currentFromLang = text.split(" ")[1];
                sendMessage(chatId, "Source language set to: " + currentFromLang);
            }
            else if (text.startsWith("/to ")) {
                currentToLang = text.split(" ")[1];
                sendMessage(chatId, "Target language set to: " + currentToLang);
            }
            else if (text.startsWith("/help")) {
                sendHelpMessage(chatId);
            }
            else if (text.startsWith("/current")) {
                String currentText = "Your current native language is set to *" + currentFromLang + "*\\.\nYour current translating language is *" + currentToLang + "*\\.\n" +
                        "To change it, write '/from en' to change 'from' language to another \\(e\\.g\\. it is en\\) and write '/to es' if you want to change 'to' lang \\(e\\.g\\. " +
                        "it is espanol\\)" +
                        "\n*In short, it is* " + currentFromLang + " \\-\\> " + currentToLang;
                sendImage(chatId, currentText);
            }
            else {
                try {

                    sendTranslatedMessageWithButtons(chatId, text, translated);
                } catch (Exception e) {
                    sendMessage(chatId, "⚠️ Error: " + e.getMessage());

                }
            }
        }
    }

    private void sendTranslatedMessageWithButtons(long chatId, String originalText, String translated) {
        InlineKeyboardMarkup navMarkup = new InlineKeyboardMarkup();
        List<List<InlineKeyboardButton>> keyboard = new ArrayList<>();

        keyboard.add(List.of(
                InlineKeyboardButton.builder().text("Original Text").callbackData(originalText).build()
        ));

        navMarkup.setKeyboard(keyboard);

        SendMessage navMessage = new SendMessage();
        navMessage.setChatId(String.valueOf(chatId));
        navMessage.setText(translated);
        navMessage.setParseMode("Markdown");
        navMessage.setReplyMarkup(navMarkup);
        try {
            execute(navMessage);
        } catch (TelegramApiException e) {
            e.printStackTrace();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

    }
    private void sendHelpMessage(long chatId) {
        String help = "📚 Translation Bot Help:\n\n" +
                "/from [lang] - Set source language (e.g. /from en)\n" +
                "/to [lang] - Set target language (e.g. /to es)\n" +
                "Just type text to translate!\n\n" +
                "Current settings: " + currentFromLang + " → " + currentToLang;

        sendMessage(chatId, help);
    }

    private void sendMessage(long chatId, String text) {
        SendMessage message = new SendMessage();
        message.setChatId(String.valueOf(chatId));
        message.setText(text);

        try {
            execute(message);
        } catch (TelegramApiException e) {
            e.printStackTrace();
        }
    }

    @Override
    public String getBotUsername() {
        return "TranslatorByShuxrinhoBot";
    }

    @Override
    public String getBotToken() {
        return "7210184880:AAGfOyS9BGkMKP-5mrFIjLaFwEM6W56P4Go";
    }

    private void sendMainMenu(long chatId) {
        SendMessage message = new SendMessage();
        message.setChatId(chatId);
        message.setText("Welcome to my Bot! Enjoy using it. " +
                "\nContact: @Shuxrinho\n\nIf you have any problem " +
                "with language abbreviations, here is the full list " +
                "for 5 supported languages: " +
                "\nUzbek - uz \n" +
                "English - en" +
                "\nEspanol - es" +
                "\nChinese - zh" +
                "\nRussian - ru");
        ReplyKeyboardMarkup keyboardMarkup = new ReplyKeyboardMarkup();
        List<KeyboardRow> keyboard = new ArrayList<>();

        // First row
        KeyboardRow row1 = new KeyboardRow();
        row1.add("/start");
        row1.add("/help");

        // Second row
        KeyboardRow row2 = new KeyboardRow();
        row2.add("/from");
        row2.add("/to");

        KeyboardRow row3 = new KeyboardRow();
        row3.add("/current");

        // Add rows to keyboard
        keyboard.add(row1);
        keyboard.add(row2);
        keyboard.add(row3);

        // Set keyboard properties
        keyboardMarkup.setKeyboard(keyboard);
        keyboardMarkup.setResizeKeyboard(true);
        keyboardMarkup.setOneTimeKeyboard(false);
        message.setReplyMarkup(keyboardMarkup);

        try {
            execute(message);
        } catch (TelegramApiException e) {
            e.printStackTrace();
        }

    }
    private void sendImage(long chatId, String text) {
        try {
            // Step 1: Get the correct direct download URL
            String fileId = "13zDSrCP5oGMqFJ8fz0r_AtoAELjk1EJB"; // From Google Drive shareable link
            String directImageUrl = "https://drive.google.com/uc?export=view&id=" + fileId;

            SendPhoto photo = new SendPhoto();
            photo.setChatId(String.valueOf(chatId));
            photo.setPhoto(new InputFile(directImageUrl));
            photo.setCaption(text);
            photo.setParseMode("MarkdownV2");

            execute(photo);
        } catch (TelegramApiException e) {
            e.printStackTrace();
            // Fallback: Download and send as file if URL doesn't work
            try {
                sendAsDocument(chatId);
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
    }
    private void sendAsDocument(long chatId) throws Exception {
        String fileId = "13zDSrCP5oGMqFJ8fz0r_AtoAELjk1EJB";
        URL downloadUrl = new URL("https://drive.google.com/uc?export=download&id=" + fileId);

        SendDocument document = new SendDocument();
        document.setChatId(String.valueOf(chatId));
        document.setDocument(new InputFile(downloadUrl.openStream(), "image.jpg"));
        execute(document);
    }

}

File: python_bot/my_bot.py
"""
Telegram Bot for translation using MyMemory API.
"""
import os
from telegram import Update, InlineKeyboardButton, ReplyKeyboardMarkup, KeyboardButton, InputFile
from telegram.ext import Application, CommandHandler, MessageHandler, filters, ContextTypes, CallbackQueryHandler
from telegram.constants import ParseMode

from my_memory_translator import MyMemoryTranslator


class MyBot:
    """A Telegram bot that translates text using MyMemory API."""
    
    BOT_TOKEN = "7210184880:AAGfOyS9BGkMKP-5mrFIjLaFwEM6W56P4Go"
    BOT_USERNAME = "TranslatorByShuxrinhoBot"
    
    def __init__(self):
        self.translator = MyMemoryTranslator()
        self.current_from_lang = "en"
        self.current_to_lang = "es"
        self.application = None
        
    async def start_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Handle /start command."""
        await self.send_main_menu(update.effective_chat.id)
        
    async def help_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Handle /help command."""
        chat_id = update.effective_chat.id
        help_text = (
            "📚 Translation Bot Help:\n\n"
            "/from [lang] - Set source language (e.g. /from en)\n"
            "/to [lang] - Set target language (e.g. /to es)\n"
            "Just type text to translate!\n\n"
            f"Current settings: {self.current_from_lang} → {self.current_to_lang}"
        )
        await context.bot.send_message(chat_id=chat_id, text=help_text)
        
    async def from_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Handle /from command to set source language."""
        chat_id = update.effective_chat.id
        if context.args and len(context.args) > 0:
            self.current_from_lang = context.args[0]
            await context.bot.send_message(
                chat_id=chat_id, 
                text=f"Source language set to: {self.current_from_lang}"
            )
        else:
            await context.bot.send_message(
                chat_id=chat_id, 
                text="Please provide a language code. Example: /from en"
            )
            
    async def to_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Handle /to command to set target language."""
        chat_id = update.effective_chat.id
        if context.args and len(context.args) > 0:
            self.current_to_lang = context.args[0]
            await context.bot.send_message(
                chat_id=chat_id, 
                text=f"Target language set to: {self.current_to_lang}"
            )
        else:
            await context.bot.send_message(
                chat_id=chat_id, 
                text="Please provide a language code. Example: /to es"
            )
            
    async def current_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Handle /current command to show current language settings."""
        chat_id = update.effective_chat.id
        current_text = (
            f"Your current native language is set to *{self.current_from_lang}*.\n"
            f"Your current translating language is *{self.current_to_lang}*.\n"
            f"To change it, write '/from en' to change 'from' language to another (e.g. it is en) "
            f"and write '/to es' if you want to change 'to' lang (e.g. it is espanol)\n"
            f"*In short, it is* {self.current_from_lang} -> {self.current_to_lang}"
        )
        await self.send_image(chat_id, current_text)
        
    async def handle_message(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Handle incoming text messages."""
        if update.message and update.message.text:
            text = update.message.text
            chat_id = update.effective_chat.id
            
            # Check for commands embedded in message
            if text.startswith("/start"):
                await self.send_main_menu(chat_id)
                return
            elif text.startswith("/from "):
                parts = text.split(" ")
                if len(parts) > 1:
                    self.current_from_lang = parts[1]
                    await context.bot.send_message(
                        chat_id=chat_id,
                        text=f"Source language set to: {self.current_from_lang}"
                    )
                return
            elif text.startswith("/to "):
                parts = text.split(" ")
                if len(parts) > 1:
                    self.current_to_lang = parts[1]
                    await context.bot.send_message(
                        chat_id=chat_id,
                        text=f"Target language set to: {self.current_to_lang}"
                    )
                return
            elif text.startswith("/help"):
                await self.help_command(update, context)
                return
            elif text.startswith("/current"):
                await self.current_command(update, context)
                return
            
            # Translate the text
            try:
                translated = self.translator.translate(
                    text, 
                    self.current_from_lang, 
                    self.current_to_lang
                )
                await self.send_translated_message_with_buttons(
                    chat_id, 
                    text, 
                    translated
                )
            except Exception as e:
                await context.bot.send_message(
                    chat_id=chat_id,
                    text=f"⚠️ Error: {str(e)}"
                )
                
    async def send_translated_message_with_buttons(self, chat_id: int, original_text: str, translated: str):
        """Send translated message with inline keyboard button."""
        keyboard = [[InlineKeyboardButton("Original Text", callback_data=original_text)]]
        
        await self.application.bot.send_message(
            chat_id=chat_id,
            text=translated,
            parse_mode=ParseMode.MARKDOWN,
            reply_markup=InlineKeyboardMarkup(keyboard)
        )
        
    async def button_callback(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Handle inline keyboard button callbacks."""
        query = update.callback_query
        if query:
            await query.answer()
            await query.edit_message_text(text=query.data)
            
    async def send_main_menu(self, chat_id: int):
        """Send main menu with reply keyboard."""
        message_text = (
            "Welcome to my Bot! Enjoy using it. "
            "\nContact: @Shuxrinho\n\nIf you have any problem "
            "with language abbreviations, here is the full list "
            "for 5 supported languages: "
            "\nUzbek - uz \n"
            "English - en"
            "\nEspanol - es"
            "\nChinese - zh"
            "\nRussian - ru"
        )
        
        keyboard = [
            [KeyboardButton("/start"), KeyboardButton("/help")],
            [KeyboardButton("/from"), KeyboardButton("/to")],
            [KeyboardButton("/current")]
        ]
        
        reply_markup = ReplyKeyboardMarkup(
            keyboard, 
            resize_keyboard=True, 
            one_time_keyboard=False
        )
        
        await self.application.bot.send_message(
            chat_id=chat_id,
            text=message_text,
            reply_markup=reply_markup
        )
        
    async def send_image(self, chat_id: int, caption: str):
        """Send image from Google Drive with caption."""
        file_id = "13zDSrCP5oGMqFJ8fz0r_AtoAELjk1EJB"
        direct_image_url = f"https://drive.google.com/uc?export=view&id={file_id}"
        
        try:
            await self.application.bot.send_photo(
                chat_id=chat_id,
                photo=direct_image_url,
                caption=caption,
                parse_mode=ParseMode.MARKDOWN_V2
            )
        except Exception as e:
            print(f"Error sending photo: {e}")
            # Fallback: send as document
            try:
                await self.send_as_document(chat_id)
            except Exception as ex:
                print(f"Error sending document: {ex}")
                
    async def send_as_document(self, chat_id: int):
        """Send image as document (fallback method)."""
        file_id = "13zDSrCP5oGMqFJ8fz0r_AtoAELjk1EJB"
        download_url = f"https://drive.google.com/uc?export=download&id={file_id}"
        
        await self.application.bot.send_document(
            chat_id=chat_id,
            document=download_url,
            filename="image.jpg"
        )
        
    def run(self):
        """Run the bot."""
        # Create application
        self.application = Application.builder().token(self.BOT_TOKEN).build()
        
        # Add handlers
        self.application.add_handler(CommandHandler("start", self.start_command))
        self.application.add_handler(CommandHandler("help", self.help_command))
        self.application.add_handler(CommandHandler("from", self.from_command))
        self.application.add_handler(CommandHandler("to", self.to_command))
        self.application.add_handler(CommandHandler("current", self.current_command))
        self.application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, self.handle_message))
        self.application.add_handler(CallbackQueryHandler(self.button_callback))
        
        # Start the bot
        print("Bot successfully started!")
        self.application.run_polling(allowed_updates=Update.ALL_TYPES)


def main():
    """Main entry point."""
    bot = MyBot()
    bot.run()


if __name__ == "__main__":
    main()
